feat: add claim-rewards-to-self functionality to incentives controller

SeanJCasey opened the pull request:
This PR adds a `claimRewardsToSelf()` function to the incentives controller that reuses the same underlying `_claimRewards()` function as other claim functions, but auto-populating the `to` field with `msg.sender`.

SeanJCasey committed:
feat: add claimRewardsToSelf() function to incentives controller contracts

SeanJCasey committed:
test: add test suite for claimRewardsToSelf()

ernesto-usal commented:
2 comments @SeanJCasey :
- The `REVISION` will need to be incremented to `2`, to follow the upgradeability requirements, here https://github.com/aave/incentives-proposal/pull/4/files#diff-666660e636b6be0369551917c2269b9104d3d24ede682a32f4b027f968738250R30
- Being the second version of the `StakedTokenIncentivesController`, the `initialize()` function can be empty, not receiving any parameter here https://github.com/aave/incentives-proposal/pull/4/files#diff-666660e636b6be0369551917c2269b9104d3d24ede682a32f4b027f968738250R55

ernesto-usal commented:
I also suggest @SeanJCasey to create mainnet fork tests reproducing a real scenario of claiming rewards from a system that will be using this function. For that you will need to simulate the upgrade of the proxy contract with the new implementation

SeanJCasey committed:
refactor: bump REVISION in incentives controller

SeanJCasey committed:
refactor: empty initialize() in incentives controller

SeanJCasey committed:
chore: bump mainnet fork block

SeanJCasey committed:
chore: fix types so selfdestruct transfer helper works

SeanJCasey committed:
test: add mainnet fork test suite for incentives revision 2

SeanJCasey committed:
chore: add script for running mainnet fork test suite

SeanJCasey commented:
@ernesto-usal I implemented the suggested contract changes and added the suggested mainnet fork test suite.

I also added a script to package.json for running the mainnet fork test suite.

I realized during this process that the current tests in `test-fork/` don't work (unless I'm mistaken). I'm guessing that they were just cloned over from the staked aave repo. Anyways, I needed to bump the fork block to get incentives working and needed to change a type that didn't exist in this repo.

Also, I wasn't sure if I needed to preserve the use of `DRE.ethers` in this file, so I included it anyways.

SeanJCasey commented:
Also, I kept the `initialize()` function, but just emptied it and removed the no-longer-necessary modifier. I assume this is what you wanted.

ernesto-usal commented:
@SeanJCasey actually the `initializer` modifier should remain, because if not, the state (on `VersionedInitializable`) concerning the versioning will not be modified, and could lead to inconsistencies on consequent updates

SeanJCasey committed:
refactor: reimplement initializer modifier

SeanJCasey committed:
docs: add audit report

SeanJCasey committed:
feat: add proposal executor contract

SeanJCasey committed:
test: add e2e test suite for proposal executor

SeanJCasey committed:
chore: add command to run proposal execution test suite

SeanJCasey committed:
chore: update hardhat dependency

SeanJCasey committed:
chore: update hardhat config global default to use london fork and update mainnet fork blocknumber

miguelmtzinf commented on contracts/proposals/ProposalIncentivesRev2Executor.sol:
```diff
@@ -0,0 +1,18 @@
+contract ProposalIncentivesRev2Executor is IProposalIncentivesRev2Executor {
+  address constant INCENTIVES_CONTROLLER_IMPL_ADDRESS = 0xD9ED413bCF58c266F95fE6BA63B13cf79299CE31;
+  address constant LENDING_POOL_ADDRESS_PROVIDER = 0xB53C1a33016B2DC2fF3653530bfF1848a515c8c5;
+  bytes32 constant LENDING_POOL_ID = bytes32(keccak256(bytes('INCENTIVES_CONTROLLER')));
```
incorrect variable name

SeanJCasey replied:
what's the correct variable name?

miguelmtzinf replied:
Its not `LENDING_POOL_ID`, but `INCENTIVES_CONTROLLER_ID`

miguelmtzinf reviewed (changes requested):
I would change the name of the audit file to include the date.

SeanJCasey committed:
test: add accrual assertion to incentives rev2 test

SeanJCasey committed:
test: add pre- and post-upgrade accrual checks

miguelmtzinf commented on contracts/incentives/StakedTokenIncentivesController.sol:
```diff
@@ -49,13 +49,9 @@ contract StakedTokenIncentivesController is
   /**
-   * @dev Initialize IStakedTokenIncentivesController
-   * @param addressesProvider the address of the corresponding addresses provider
+   * @dev Initialize IStakedTokenIncentivesController. Empty after REVISION 1, but maintains the expected interface.
```
Second sentence a bit confusing (there is no interface). Would remove it.

miguelmtzinf commented on contracts/interfaces/IProposalIncentivesRev2Executor.sol:
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity 0.7.5;
+
+interface IProposalIncentivesRev2Executor {
```
`IProposalIncentivesV2Executor` name feels more aligned imo

miguelmtzinf commented on test-fork/incentives-revision-2-upgrade.spec.ts:
```diff
@@ -0,0 +1,194 @@
+    expect(preUpgradeRewardsBalance).to.be.gt(0);
+  });
+
+  it('upgrades', async () => {
```
In order to test the whole process, would be nice to run the gov proposal process here (no impersonating).
Then, I think would be nice to move the process of the other test file here.

SeanJCasey committed:
docs: update audit filename

SeanJCasey committed:
refactor: fix variable name

SeanJCasey committed:
refactor: update contract namespace

SeanJCasey committed:
test: update test suite to use actual gov proposal process

SeanJCasey committed:
test: update test names to match new contract namespace

miguelmtzinf commented on hardhat.config.ts:
```diff
@@ -58,7 +58,7 @@ const getCommonNetworkConfig = (networkName: eNetwork, networkId: number) => ({
 const mainnetFork = MAINNET_FORK
   ? {
-      blockNumber: 12290275,
+      blockNumber: 13410750, // Oct 13, 2021
```
Lets pass here an env variable, so its possible to adjust its value depending on the task/test.
```
blockNumber: parseInt(process.env.FORKING_BLOCK)
```

miguelmtzinf commented on package.json:
```diff
@@ -19,7 +19,9 @@
     "coverage": "hardhat coverage",
     "test": "npm run test-incentives",
     "test-incentives": "TS_NODE_TRANSPILE_ONLY=1 hardhat test ./test/__setup.spec.ts ./test/AaveIncentivesController/*.spec.ts",
```
Lets add `FORKING_BLOCK=12290275` for the previous tests to make them work.

miguelmtzinf commented on package.json:
```diff
@@ -19,7 +19,9 @@
     "coverage": "hardhat coverage",
     "test": "npm run test-incentives",
     "test-incentives": "TS_NODE_TRANSPILE_ONLY=1 hardhat test ./test/__setup.spec.ts ./test/AaveIncentivesController/*.spec.ts",
+    "test-incentives-upgrade-revision-2": "TS_NODE_TRANSPILE_ONLY=1 MAINNET_FORK=true hardhat test ./test-fork/incentives-revision-2-upgrade.spec.ts",
```
The name of the file seems wrong.

SeanJCasey replied:
i think this was already addressed in the prev commit, but lmk if not the case

miguelmtzinf commented on package.json:
```diff
@@ -19,7 +19,9 @@
     "test-incentives": "TS_NODE_TRANSPILE_ONLY=1 hardhat test ./test/__setup.spec.ts ./test/AaveIncentivesController/*.spec.ts",
+    "test-incentives-upgrade-revision-2": "TS_NODE_TRANSPILE_ONLY=1 MAINNET_FORK=true hardhat test ./test-fork/incentives-revision-2-upgrade.spec.ts",
     "test-proposal": "TS_NODE_TRANSPILE_ONLY=1 MAINNET_FORK=true hardhat test ./test-fork/incentivesProposal.spec.ts",
+    "test-proposal-rev2": "TS_NODE_TRANSPILE_ONLY=1 MAINNET_FORK=true hardhat test ./test-fork/incentivesProposalRev2.spec.ts",
```
Same as above

SeanJCasey replied:
same

SeanJCasey committed:
chore: pass in blocknumber as env var for fork tests

SeanJCasey committed:
chore: grant max allowance of AAVE to stkAAVE for incentives controller during unit test setup

deauna commented:
@0x28c04471ce4f5d1f027ca5f69faa5678c6f87937
